libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -262,63 +275,74 @@ void ContractLevelChecker::checkIllegalOverrides(ContractDefinition const& _cont
+template<class T>
+bool ContractLevelChecker::checkFunctionOverride(T const& _overriding, FunctionDefinition const& _super)
 {
-	FunctionTypePointer functionType = FunctionType(_function).asCallableFunction(false);
+	string baseTypeName;
+
+	if constexpr(std::is_same<FunctionDefinition, T>::value)
+		baseTypeName = "function";
+	else
+		baseTypeName = "public state variable";
+
+	FunctionTypePointer functionType = FunctionType(_overriding).asCallableFunction(false);
 	FunctionTypePointer superType = FunctionType(_super).asCallableFunction(false);
 
 	bool success = true;
 
 	if (!functionType->hasEqualParameterTypes(*superType))
 		return true;
 
-	if (!_function.overrides())
+	if (!_overriding.overrides())
 	{
-		overrideError(_function, _super, "Overriding function is missing 'override' specifier.");
+		overrideError(_overriding, _super, "Overriding " + baseTypeName + " is missing 'override' specifier.");
 		success = false;
 	}
 
 	if (!_super.virtualSemantics())
 	{
-		overrideError( _super, _function, "Trying to override non-virtual function. Did you forget to add \"virtual\"?", "Overriding function is here:");
+		overrideError( _super, _overriding, "Trying to override non-virtual function. Did you forget to add \"virtual\"?", "Overriding " + baseTypeName + " is here:");
 		success = false;
 	}
 
 	if (!functionType->hasEqualReturnTypes(*superType))
 	{
-		overrideError(_function, _super, "Overriding function return types differ.");
+		overrideError(_overriding, _super, "Overriding " + baseTypeName + " return types differ.");
 		success = false;
 	}
 
-	if (!_function.annotation().superFunction)
-		_function.annotation().superFunction = &_super;
+	if constexpr(std::is_same<T, FunctionDefinition>::value)
+		if (!_overriding.annotation().superFunction)
+			_overriding.annotation().superFunction = &_super;
 
-	if (_function.visibility() != _super.visibility())
+	if (_overriding.visibility() != _super.visibility())
 	{
 		// Visibility change from external to public is fine.
 		// Any other change is disallowed.
 		if (!(
 			_super.visibility() == FunctionDefinition::Visibility::External &&
-			_function.visibility() == FunctionDefinition::Visibility::Public
+			_overriding.visibility() == FunctionDefinition::Visibility::Public
 		))
 		{
-			overrideError(_function, _super, "Overriding function visibility differs.");
+			overrideError(_overriding, _super, "Overriding " + baseTypeName + " visibility differs.");
 			success = false;
 		}
 	}
-	if (_function.stateMutability() != _super.stateMutability())
-	{
-		overrideError(
-			_function,
-			_super,
-			"Overriding function changes state mutability from \"" +
-			stateMutabilityToString(_super.stateMutability()) +
-			"\" to \"" +
-			stateMutabilityToString(_function.stateMutability()) +
-			"\"."
-		);
-		success = false;
-	}
+
+	if constexpr(std::is_same<T, FunctionDefinition>::value)
+		if (_overriding.stateMutability() != _super.stateMutability())
```

Comment:
I wonder whether we should take the feature recently brought to light by @fulldecent still into 0.6.0: Allowing to soften the mutability requirements when overriding. It might get complicated because mutability is not a linear order and also when taking multiple inheritance into account, but maybe not...